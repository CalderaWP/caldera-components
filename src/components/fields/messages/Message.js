import React from 'react';
import PropTypes from 'prop-types';
import {messagePropShape} from "./messagePropTypes";
import classNames from 'classnames';

/**
 * The class that message components always wrap messages in
 * @type {string}
 */
export const MESSAGE_CLASS = 'caldera-components-message';

/**
 * Prop for showing a validation message
 *
 * Use or for errors and such
 *
 * @param {object}  props
 * @return {*}
 * @constructor
 */
export const Message = (props) => {
	const {message,error} = props.message;
	if( !message || '' === message ){
		return <React.Fragment></React.Fragment>;
	}
	return(
		<div
			className={classNames(
				props.className,
				MESSAGE_CLASS,
				{
					'has-error': error,
					'caldera-components-error': error,
					'caldera-components-not-error': ! error,

				}
			)}
		>
			{message}
		</div>
	)
};

/**
 * Prop types for Message components
 *
 * @type {{className: shim, message}}
 */
Message.propTypes = {
	className: PropTypes.string,
	message: messagePropShape,
};

