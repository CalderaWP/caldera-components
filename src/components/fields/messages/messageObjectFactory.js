import {toBoolean} from '../util';

/**
 * Prepares messages object in fieldConfigs
 *
 * @param {Object} message
 * @param {Boolean} error
 * @return {{message: *, error: boolean}}
 */
export const messageObjectFactory = ({message,error}) => {
	if( 'string' !== typeof  message ){
		message = '';
	}
	return{
		message: message,
		error: toBoolean(error)
	};
};