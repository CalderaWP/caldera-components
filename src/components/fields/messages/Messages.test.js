import renderer from 'react-test-renderer';
import React from 'react';
import { shallow } from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Message, MESSAGE_CLASS} from './Message';
import {messageObjectFactory} from './messageObjectFactory';

Enzyme.configure({ adapter: new Adapter() });

describe( 'Message component', () => {
	describe( 'Classnames', () => {
		describe( 'Displaying message', () => {
			it( 'Shows a message', () => {
				const wrapper = shallow(
					<Message
						message={{
							message:'Hi Roy'
						}}
					/>
				);
				expect( wrapper ).toHaveLength(1);
			});

			it( 'Renders message', () => {
				const wrapper = renderer.create(
					<Message
						message={{
							message:'Hi Roy',
							error: false,
						}}
					/>
				);
				expect( wrapper.toJSON() ).toMatchSnapshot();
			});

			it( 'Shows the right message message', () => {
				const wrapper = shallow(
					<Message
						message={{
							message:'Hi Roy'
						}}
					/>
				);
				expect( wrapper.text() ).toBe('Hi Roy');
			});

		});

		it( 'shows classname prop as class', () => {
			const wrapper = shallow(
				<Message
					message={{
						message:'Hi Roy'
					}}
					className={'mike'}
				/>
			);
			expect( wrapper.find( '.mike') ).toHaveLength(1);
		});

		it( 'Has error class when is an error', () => {
			const wrapper = shallow(
				<Message
					message={{
						message:'Something bad has happened.',
						error: true,
					}}
				/>
			);
			expect( wrapper.find( '.has-error') ).toHaveLength(1);
		});

		it( 'Does not have error class when is not an error', () => {
			const wrapper = shallow(
				<Message
					message={{
						message:'Something good has happened.',
						error: false,
					}}
				/>
			);
			expect( wrapper.find( '.has-error') ).toHaveLength(0);
		});

		it( 'Does have caldera-components-error class when an error', () => {
			const wrapper = shallow(
				<Message
					message={{
						message:'Something',
						error: true,
					}}
				/>
			);
			expect( wrapper.hasClass( MESSAGE_CLASS ) ).toEqual(true);
		});

		it( 'Does have caldera-components-error class when not an error', () => {
			const wrapper = shallow(
				<Message
					message={{
						message:'Something',
						error: false,
					}}
				/>
			);
			expect( wrapper.hasClass( MESSAGE_CLASS ) ).toEqual(true);
		});
	});

});

describe( 'messageObjectFactory', () => {
	it( 'casts truthy error to boolean', () => {
		expect( messageObjectFactory({
			error: 1,
			message: 'Roy'
		}).error).toBe(true);
	});

	it( 'casts falsey error to boolean', () => {
		expect( messageObjectFactory({
			error: 'false',
			message: 'Roy'
		}).error).toBe(false);
	});
});