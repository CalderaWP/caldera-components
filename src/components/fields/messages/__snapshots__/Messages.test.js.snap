// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component Classnames Displaying message Renders message 1`] = `
<div
  className="caldera-components-message caldera-components-not-error"
>
  Hi Roy
</div>
`;
